crates/kornia-vlm/src/smolvlm/mod.rs

```diff
@@ -0,0 +1,277 @@
+    if let Some((Width(w), _)) = terminal_size() {
+        (text.len() + w as usize + 1) / w as usize
+    } else {
+        1
+    }
+}
+
+fn clear_lines(n: usize) {
+    for _ in 0..n {
+        print!("\x1B[1A\x1B[2K");
+    }
+    io::stdout().flush().unwrap();
+}
+
+fn read_input(cli_prompt: &str) -> String {
+    let mut input = String::new();
+    print!("{}", cli_prompt);
+    io::stdout().flush().unwrap();
+    io::stdin()
+        .read_line(&mut input)
+        .expect("Failed to read input");
+
+    input.trim().to_owned()
+}
+
+#[derive(thiserror::Error, Debug)]
+pub enum SmolVlmError {
+    #[error(transparent)]
+    FailedToLoadModel(#[from] hf_hub::api::sync::ApiError),
+
+    #[error(transparent)]
+    CandleError(#[from] candle_core::Error),
+
+    #[error(transparent)]
+    ImageError(#[from] kornia_image::ImageError),
+
+    #[error(transparent)]
+    TokenizerError(#[from] tokenizers::Error),
+
+    #[error(transparent)]
+    IoError(#[from] std::io::Error),
+
+    // TODO: not used right now (currently, the end token is handled via if/else, which might be preferred)
+    #[error("Cannot find the <end_of_utterance> token")]
+    EosTokenNotFound,
+}
+
+pub struct SmolVlm {
+    model: text_model::SmolModel,
+    tokenizer: Tokenizer,
+    image_token_enc: Encoding,
+    config: SmolVlmConfig,
+    rng: StdRng,
+    device: Device,
+}
+
+impl SmolVlm {
+    /// Create a new SmolVlm model
+    ///
+    /// # Arguments
+    ///
+    /// * `config` - The configuration for the SmolVlm model
+    ///
+    /// # Returns
+    pub fn new(config: SmolVlmConfig) -> Result<Self, SmolVlmError> {
+        #[cfg(feature = "cuda")]
+        let (device, dtype) = match Device::cuda_if_available(0) {
+            Ok(device) => (device, DType::BF16),
+            Err(e) => {
+                log::warn!("CUDA not available, defaulting to CPU: {}", e);
+                (Device::Cpu, DType::F32)
+            }
+        };
+
+        #[cfg(not(feature = "cuda"))]
+        let (device, dtype) = (Device::Cpu, DType::F32);
+
+        // TODO: find a way to use FP32 if cuda is not available
+
+        let (model, tokenizer) = Self::load_model(dtype, &device)?;
+        let image_token_enc = tokenizer.encode("<image>", false)?;
+        let rng = StdRng::seed_from_u64(config.seed);
+
+        Ok(Self {
+            model,
+            tokenizer,
+            image_token_enc,
+            config,
+            device,
+            rng,
+        })
+    }
+
+    /// Run the inference of the SmolVLM model
+    ///
+    /// # Arguments
+    ///
+    /// * `image` - The rgb8    image to generate a caption for with shape [H, W, 3]
+    /// * `prompt` - The prompt to generate a caption for
+    /// * `sample_len` - The length of the generated caption
+    /// * `stdout_debug` - Whether to print the debug information to the stdout
+    ///
+    /// # Returns
+    ///
+    /// * `caption` - The generated caption
+    pub fn inference(
```

Comment:
mimic this api: https://github.com/kornia/kornia-rs/blob/main/crates/kornia-vlm/src/paligemma/mod.rs#L127-L133
so that we can pass in memory images to build other applications on top of it